### ISSUE TITLE
Macro invocations are not expanded in the `if` expressions

### ISSUE THREAD

**Issue by antego** [Rust-GCC/gccrs#1141]
Related to the #1119 

Looks like the macro expansion works incorrectly in the comparison expressions and in the `if` expressions

I tried this code:

```rust
  let res = "test1" == "test1";
  if res {
    print(0);
  }
  let res = (concat!("test2") == "test3");
  if res {
    print(1);
  }
  if concat!("test2") == "test3" {
    print(2);
  }
```

I expected the code to compile successfully. Instead, the compilation failed with the error:
```
fatal error: Failed to lower expr: [MacroInvocation:
 outer attributes: none
 concat!(("test2"))
 has semicolon: false]
compilation terminated.
```

### Meta

From the zulip chat:

```
@antego:
Some conclusions from the ast above:

macros are expanded outside of the if expressions, but they remove all other parts of the expression. concat!("test2") == "test3" became just test2
macros are not expanded inside the if expressions, leading to the error

Questions:

Are the above conclusions correct?
Any hints about where to look next? I was planning to go through the code and find the place where the macro expansion is called and check how the expanded macro is handled for a simple expression vs if expression.

@CohenArthur:
Macro expansion is done in our attribute visitor, which you can find the implementation of in gcc/rust/expand/[rust-attribute-visitor.cc](http://rust-attribute-visitor.cc/). The code is a little bit hard to follow as it does a lof of in-place replacements of the expanded AST fragments (btw, we could probably think about splitting the attribute visitor and in-place expansion phase. Even though it might be a slowdown?).

The reason it crashes and you get the "Couldn't lower expression" error is that the macro fragment simply does not get used in the proper place. You can look at the visitor for AST::LetExpr values and see how you are supposed to expand an expression in-place.

I've checked the code, and it hasn't been implemented for all AST::If{Expr, LetExpr...} nodes yet. You can probably do the same as a LetExpr and add tests to see your progress gradually.

So to answer your questions precisely:

a/ Yes and no! These are probably side-effects of the code just not existing yet. Sorry about that!
b/ Yes! That's completely right

The rust-attribute-visitor.cc file is a good place to look at, and specifically LetExpr or ArithmeticExpr which do expression expansion :)
```

**Comment by liushuyu** [comment]
Probably related: #1084 and #1048

**Comment by CohenArthur** [comment]
Fixed by #1170
